File: ios/DouyinDownLoad/Services/CookieStore.swift
import Foundation

/// 管理各平台的认证 Cookie（持久化到 UserDefaults）
/// nonisolated 以便从任意隔离域访问（UserDefaults 自身线程安全）
nonisolated final class CookieStore: @unchecked Sendable {
    static let shared = CookieStore()

    private let defaults = UserDefaults.standard
    private let storageKey = "platform_cookies"

    private init() {}

    // MARK: - 平台定义

    struct PlatformCookieConfig {
        let platform: String
        let displayName: String
        let fields: [(key: String, label: String, placeholder: String)]
    }

    static let supportedPlatforms: [PlatformCookieConfig] = [
        PlatformCookieConfig(
            platform: "instagram",
            displayName: "Instagram",
            fields: [
                (key: "sessionid", label: "sessionid", placeholder: "粘贴 sessionid 值"),
                (key: "ds_user_id", label: "ds_user_id", placeholder: "粘贴 ds_user_id 值"),
                (key: "csrftoken", label: "csrftoken", placeholder: "粘贴 csrftoken 值"),
            ]
        ),
        PlatformCookieConfig(
            platform: "x",
            displayName: "X (Twitter)",
            fields: [
                (key: "auth_token", label: "auth_token", placeholder: "粘贴 auth_token 值"),
                (key: "ct0", label: "ct0", placeholder: "粘贴 ct0 值"),
            ]
        ),
    ]

    // MARK: - 读写

    /// 获取所有平台的 cookies，用于发送给 backend
    func allCookies() -> [String: [String: String]] {
        guard let data = defaults.data(forKey: storageKey),
              let dict = try? JSONDecoder().decode([String: [String: String]].self, from: data) else {
            return [:]
        }
        // 过滤掉空值
        return dict.compactMapValues { cookies in
            let filtered = cookies.filter { !$0.value.isEmpty }
            return filtered.isEmpty ? nil : filtered
        }.compactMapValues { $0 }
    }

    /// 获取某平台的 cookies
    func cookies(for platform: String) -> [String: String] {
        allCookies()[platform] ?? [:]
    }

    /// 保存某平台的 cookies
    func saveCookies(for platform: String, cookies: [String: String]) {
        var all = allCookies()
        let filtered = cookies.filter { !$0.value.isEmpty }
        if filtered.isEmpty {
            all.removeValue(forKey: platform)
        } else {
            all[platform] = filtered
        }
        if let data = try? JSONEncoder().encode(all) {
            defaults.set(data, forKey: storageKey)
        }
    }

    /// 清除某平台的 cookies
    func clearCookies(for platform: String) {
        var all = allCookies()
        all.removeValue(forKey: platform)
        if let data = try? JSONEncoder().encode(all) {
            defaults.set(data, forKey: storageKey)
        }
    }

    /// 检查某平台是否已配置 cookie
    func hasCookies(for platform: String) -> Bool {
        let c = cookies(for: platform)
        return !c.isEmpty && c.values.contains(where: { !$0.isEmpty })
    }
}

File: android/app/src/main/java/com/demo/videopick/data/repository/CookieStore.kt
package com.demo.videopick.data.repository

import android.content.Context
import android.content.SharedPreferences
import kotlinx.serialization.json.Json
import kotlinx.serialization.encodeToString
import kotlinx.serialization.decodeFromString

data class PlatformCookieConfig(
    val platform: String,
    val displayName: String,
    val fields: List<CookieField>,
)

data class CookieField(
    val key: String,
    val label: String,
    val placeholder: String,
)

object CookieStore {
    private const val PREFS_NAME = "platform_cookies"
    private const val KEY = "cookies_json"

    val supportedPlatforms = listOf(
        PlatformCookieConfig(
            platform = "instagram",
            displayName = "Instagram",
            fields = listOf(
                CookieField("sessionid", "sessionid", "粘贴 sessionid 值"),
                CookieField("ds_user_id", "ds_user_id", "粘贴 ds_user_id 值"),
                CookieField("csrftoken", "csrftoken", "粘贴 csrftoken 值"),
            ),
        ),
        PlatformCookieConfig(
            platform = "x",
            displayName = "X (Twitter)",
            fields = listOf(
                CookieField("auth_token", "auth_token", "粘贴 auth_token 值"),
                CookieField("ct0", "ct0", "粘贴 ct0 值"),
            ),
        ),
    )

    private fun prefs(context: Context): SharedPreferences =
        context.getSharedPreferences(PREFS_NAME, Context.MODE_PRIVATE)

    fun allCookies(context: Context): Map<String, Map<String, String>> {
        val raw = prefs(context).getString(KEY, null) ?: return emptyMap()
        return try {
            val all: Map<String, Map<String, String>> = Json.decodeFromString(raw)
            all.mapValues { (_, v) -> v.filterValues { it.isNotBlank() } }
                .filterValues { it.isNotEmpty() }
        } catch (_: Exception) {
            emptyMap()
        }
    }

    fun getCookies(context: Context, platform: String): Map<String, String> =
        allCookies(context)[platform] ?: emptyMap()

    fun saveCookies(context: Context, platform: String, cookies: Map<String, String>) {
        val all = allCookies(context).toMutableMap()
        val filtered = cookies.filterValues { it.isNotBlank() }
        if (filtered.isEmpty()) {
            all.remove(platform)
        } else {
            all[platform] = filtered
        }
        prefs(context).edit().putString(KEY, Json.encodeToString(all)).apply()
    }

    fun clearCookies(context: Context, platform: String) {
        val all = allCookies(context).toMutableMap()
        all.remove(platform)
        prefs(context).edit().putString(KEY, Json.encodeToString(all)).apply()
    }

    fun hasCookies(context: Context, platform: String): Boolean =
        getCookies(context, platform).isNotEmpty()

    fun footerText(platform: String): String = when (platform) {
        "instagram" -> "在浏览器登录 Instagram 后，使用 Cookie 导出插件获取 sessionid 等值"
        "x" -> "在浏览器登录 X/Twitter 后，使用 Cookie 导出插件获取 auth_token 和 ct0 值"
        else -> ""
    }
}
